Redesign Audio Plus Popout to look like Tumblr's Docked Video Popouts

motackt opened the pull request:
Ok, so this PR updates the Audio Plus popout to look more like Tumblr's own docked video popouts, plus adds some other features:
* Can move the progress bar around just like in the post itself! The post will update along with it
* It adjusts it's height when there's a docked video popout, so both are visible at all times

Hope it looks good! The docked audio post colour in the screenshots match my theme, but the committed .css uses Tumblr's standard white.

![audio plus docked](https://user-images.githubusercontent.com/42635772/49742274-47b00700-fcec-11e8-9666-1f5589b68d81.png)
![audio plus docked 2](https://user-images.githubusercontent.com/42635772/49742275-4979ca80-fcec-11e8-9300-d3e90972736c.png)

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -178,6 +294,34 @@ XKit.extensions.audio_plus = {
+
+		var docked_video_height = docked_video.style.height;
+		var controls_style = window.getComputedStyle(audio_plus.pop_out_controls);
+		audio_plus.pop_out_controls.style.transform = "translateY(calc(-" + docked_video_height + " - " + controls_style.bottom + "))";
```
so many operands both in this string and constructing it!! much nicer as a string literal:
```js
`translateY(calc(-${docked_video_height} - ${controls_style.bottom}))`
```

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -178,6 +294,34 @@ XKit.extensions.audio_plus = {
+	},
+
 	check_pop_out: function() {
 		var audio_plus = XKit.extensions.audio_plus;
```
i see where you got the idea now; if this function isn't used as an argument anywhere it'd be great to change this to `this` as well

motackt commented:
Do the changes look good?

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -74,13 +69,38 @@ XKit.extensions.audio_plus = {
 			};
-			var observer_dock = new MutationObserver(callback);
-			observer_dock.observe(targetNode, config);
+			this.observer_dock = new MutationObserver(callback);
```
should be defined on the extension object so the `.disconnect()` never references an undefined key (breaks extension disabling/uninstalling in some cases)

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -341,16 +326,16 @@ XKit.extensions.audio_plus = {
 		};
-		var observer_progress = new MutationObserver(callback);
-		observer_progress.observe(targetNode, config);
+		audio_plus.observer_progress = new MutationObserver(callback);
```
ibid

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -106,56 +126,54 @@ XKit.extensions.audio_plus = {
-		var audio_image = document.createElement("div");
-		audio_image.classList.add("audio-image");
+		const controls_markup = `
+			<div class=${cl.controls}>
```
would be nicer if you used `<div class="${cl.controls}">` etc rather than putting each of your strings in quotes. they're part of the HTML, not the classes, so putting them in the class variables is kinda confusing

nightpool replied:
why are we interpolating these variables at all if they're constant? IMO this should be:

```

<div class="xkit-audio-plus-controls audio-player">
    <div class="progress"></div>
        <div class="play-pause">
           <i class="icon icon_pause"></i>
    </div>
    <div class="audio-info">
        <div class="track-name"></div>
        <div class="audio-image"></div>
    </div>
</div>
<div id="xkit-audio-plus-controls-undock-container">
    <div id="xkit-audio-plus-controls-undock"></div>
</div>
```

I'm also not super happy about the mix of class-based and id-based scoping. we should use one or the other—and there's basically no reason to use `id` in modern web design

motackt replied:
i agree the interpolation is a bit redundant, but i don't understand why a mix of class and id is bad :? to clarify though, i basically followed the html that this extension had in 0.5.0

AprilSylph commented:
I notice the popout pause button doesn't turn back into a play button when the track ends - maybe that's something you could fix here while you're at it? Otherwise, this is looking really good!

AprilSylph commented:
Another nice touch to this would be adding the `.has_docked_post` class to `#right_column` when there's docked audio, but you'd have to be careful to not remove it when closing an audio dock with a video docked too. If you don't feel like doing that here that's fine - I'll do it sometime later.

And another unrelated thing I just noticed - you should bump the version to 0.6.0 or 1.0.0! We use [semantic versioning](https://semver.org/) (major.minor.patch) and given that this is a pretty big feature overhaul it deserves the recognition of at least a minor increment. I'm not that familiar with this extension's code but it definitely looks like you've fundamentally changed how the popout works, so a major increment would definitely be warranted.

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -196,8 +325,28 @@ XKit.extensions.audio_plus = {
+
 		audio_plus.current_player = player;
 		audio_plus.pop_out_controls.classList.add("showing");
+		document.body.classList.add("xkit_audio_plus_popout_showing");
```
should be removed if unused

motackt commented:
Ok! Another patch served up ^__^

Did the following this time:
- Version number bump (thanks April!)
- Resolved all requested changes
- When the track ends, the pause button turns back into a play button
- Fixed a bug where undocking the post messes up which play/pause icon is on the docked post if you redock it
- Fixed a bug where playing or pausing from the original post won't update the play/pause icon on the docked post
- Added .has_docked_audio to #right_column when the audio post is docked and vice versa (if this doesn't do what u wanted April, lemme know and I'll fix it)

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -346,10 +334,44 @@ XKit.extensions.audio_plus = {
+		}
+	}),
+
+	icon_observer: new MutationObserver(function(mutations, observer) {
```
i don't think this needs a new mutationobserver - there's already logic a few lines above to do something when the track ends, just extend that?

motackt replied:
Unfortunately the `progress_observer` only checks the progress element's width on the original audio post, it doesn't check for the icon element's classes. I don't think there's a way to check if the icon changed on the original post (by the user pausing/playing the track) using the progress element? But if I missed something please let me know!

AprilSylph replied:
ah, i didn't consider listening for that at all! nice touch!

motackt replied:
Thanks! ^__^

motackt commented on Extensions/audio_plus.js:
```diff
@@ -222,11 +204,14 @@ XKit.extensions.audio_plus = {
-		document.body.classList.remove("xkit_audio_plus_popout_showing");
 		audio_plus.current_player = null;
+		if (audio_plus.can_see_docked_posts && $("#right_column").classList.contains("has_docked_audio")) {
+			$("#right_column").classList.remove("has_docked_audio");
```
Ignore the commit I just pushed (it was premature again >__>), I'm having some trouble with 1.

I've written the logic for `"has_docked_posts` and it works so there will always be this class when there is at least one docked post (be it video or audio) and the class will disappear when there are absolutely no docked posts.

But, this messes with the `dock_observer` bc it only fires when `#right_column` classes change, and when there's a docked video post, then the user docks an audio post, the class doesn't change and so they occupy the same space.

I would suggest `"has_docked_audio" again, but achieve the dimming through a change in audio_plus.css, what do you think?

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,6 +33,10 @@
 	border-radius: 6px;
 }
 
+.right_column.has_docked_audio {
```
clearer to use the id here - while `#right_column` does also have a `.right_column` class it's confusing to (imo) imply that there could be more than one

plus we're using the id to add/remove the classes so the discrepancy is just kinda weird

motackt replied:
not a problem ^__^

motackt commented on Extensions/audio_plus.js:
```diff
@@ -50,54 +52,166 @@ XKit.extensions.audio_plus = {
+
+		//keep tabs on whether there's a docked video post
+		if (this.can_see_docked_posts) {
+			this.video_is_docked = false;
```
yeah this was a relic from trying to figure out the `.has_docked_post` logic, no use for it now ^__^

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,102 @@
+	z-index: 99999;
+}
+
+.xkit-audio-plus-pseudo-post.showing.video_is_docked {
```
is this also unused now? :P

motackt replied:
fhasdifpoashfi yup XD

AprilSylph replied:
i meant the entire rule! just removing `.video_is_docked` makes the audio popout fly up for no reason

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -50,54 +52,162 @@ XKit.extensions.audio_plus = {
-		// Create play and pause buttons modeled after the built-in ones
-		var playPause = document.createElement("div");
-		playPause.classList.add("play-pause");
+	dock_observer: new MutationObserver(function(mutations, observer) {
```
you can slim down these lines by using arrow function syntax
```js
dock_observer: new MutationObserver((mutations, observer) => {
```
since they're anonymous functions and they don't use their own `this`. furthermore, the `observer` argument is unused, meaning you can remove it and the parenthesis around the now-single argument:
```js
dock_observer: new MutationObserver(mutations => {
```

motackt replied:
Didn't know I could do that! Looks much nicer, thanks ^__^

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -183,24 +294,73 @@ XKit.extensions.audio_plus = {
+		$("#right_column").addClass("has_docked_audio");
 	},
 
+	progress_observer: new MutationObserver(function(mutations, observer) {
```
\*

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -183,24 +294,73 @@ XKit.extensions.audio_plus = {
+		}
+	}),
+
+	icon_observer: new MutationObserver(function(mutations, observer) {
```
\*

motackt commented on Extensions/audio_plus.js:
```diff
@@ -183,24 +294,73 @@ XKit.extensions.audio_plus = {
-		audio_plus.pop_out_controls.classList.add("showing");
-		audio_plus.pop_out_controls.classList.add("playing");
+		//show progress in popout container
+		var targetNode = player.querySelector(".progress");
```
I'll change it to `progress` then, thanks for explaining!

motackt replied:
hmm i use `progress` as a variable earlier in the code, how's `progress_node`? It's describes what it is without being too long, right?

motackt replied:
hjfiosdhfpioshda when i saw u +1'd my previous comment, i thought about what would make `progress` ok and kinda had a face-palm moment XD the office building metaphor actually makes a lot of sense, thanks!

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,98 @@
+	left: calc(50% + 178px);
+	bottom: 15px;
+	opacity: 0;
+	z-index: -99999;
```
`-1` does the same job

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,98 @@
+.xkit-audio-plus-pseudo-post.showing {
+	bottom: 20px;
+	opacity: 1;
+	z-index: 99999;
```
`78` would be good here (the video dock is 79 and i think the audio popout being obscured for a second is better than the video being obscured while it docks)

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,98 @@
+	min-height: unset;
+	padding: 5px;
+	bottom: 0px;
+	z-index: 10000;
```
fine as `1` (matches docked video)

AprilSylph commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,98 @@
-	height: 21px;
-	width: 21px;
 	color: #fff;
+	z-index: 99999;
```
fine as `2` (matches video undock button)

AprilSylph commented on Extensions/audio_plus.js:
```diff
@@ -181,26 +292,74 @@ XKit.extensions.audio_plus = {
+		this.current_player = player;
+		this.pop_out_controls.classList.add("showing");
+		this.pop_out_controls.classList.add("playing");
+		var ppIcon = this.pop_out_controls.querySelector('.play-pause').querySelector('.icon');
```
could just use `".play-pause .icon"` in a single selector?

motackt replied:
Just noticed this happened in 2 other places as well, I'll fix them and re-commit ^__^

motackt commented on Extensions/audio_plus.css:
```diff
@@ -33,36 +33,103 @@
 
-	z-index: 10000;
-	width: 85px;
+#search_actions_search .xkit-audio-plus-pseudo-post {
```
the id for the page u get when you search smth from the header

when you set up a list view (instead of the brick view) you get smth that looks like the dashboard, but is slightly to the right, so this rule just keeps the spacing consistent ^__^

motackt commented:
I'll keep that in mind for the next extension I write ;P
